src/IECoreMaya/SceneShapeSubSceneOverride.cpp

```diff
@@ -105,89 +105,36 @@ size_t memoryLimit()
 	{
-		// Store original indices
-		m_originalIndices = meshPrimitive->vertexIds();
-		const std::vector<int> &originalIndicesReadable = m_originalIndices->readable();
-
-		// Store triangulated indices
-		IECoreScene::MeshPrimitivePtr triangulated = IECore::runTimeCast<IECoreScene::MeshPrimitive>( meshPrimitive->copy() );
-		IntVectorDataPtr sequentialIndices( new IntVectorData() );
-		std::vector<int> &sequentialIndicesWritable =  sequentialIndices->writable();
-
-		sequentialIndicesWritable.resize( triangulated->variableSize(PrimitiveVariable::Interpolation::FaceVarying) );
-		tbb::parallel_for(
-			tbb::blocked_range<size_t>( 0, sequentialIndicesWritable.size() ), [&sequentialIndicesWritable]( const tbb::blocked_range<size_t> &r )
-			{
-				for( size_t i = r.begin(); i != r.end(); ++i )
-				{
-					sequentialIndicesWritable[i] = i;
-				}
-			}
-		);
-		// We store these indices on the mesh that we'll triangulate in order to get
-		// a mapping from new to old indices for facevarying data.
-		triangulated->variables["_indexMap"] = PrimitiveVariable( PrimitiveVariable::FaceVarying, sequentialIndices );
-
-		triangulated = MeshAlgo::triangulate( triangulated.get() );
-
-		m_triangulatedIndices = triangulated->vertexIds();
-		const std::vector<int> &triangulatedIndicesReadable = m_triangulatedIndices->readable();
-
-		std::vector<int> &oldToTriangulatedIndexMappingWritable = m_oldToTriangulatedIndexMapping->writable();
-		oldToTriangulatedIndexMappingWritable.resize( parallelMaxElement( originalIndicesReadable ) + 1 );
-
-		std::vector<int> &indicesWritable = m_indices->writable();
-		indicesWritable.resize( triangulatedIndicesReadable.size() );
-
-		tbb::parallel_for(
-			tbb::blocked_range<size_t>( 0, indicesWritable.size() ), [&indicesWritable, &triangulatedIndicesReadable, &oldToTriangulatedIndexMappingWritable]( const tbb::blocked_range<size_t> &r )
-			{
-				for( size_t i = r.begin(); i != r.end(); ++i )
-				{
-					indicesWritable[i] = i;
-					oldToTriangulatedIndexMappingWritable[ triangulatedIndicesReadable[i] ] = i;
-				}
-			}
-		);
-
-		m_mapToOldFacevarying = triangulated->expandedVariableData<IntVectorData>( "_indexMap" );
-
+		if( !primitive )
+		{
+			// null primitives represent a bounding box, which have constant topology
+			fillBoundData();
+		}
+		else if( const auto *mesh = runTimeCast<const MeshPrimitive>( primitive ) )
+		{
+			fillMeshData( mesh );
+		}
+		else if( const auto *curves = runTimeCast<const CurvesPrimitive>( primitive ) )
+		{
+			fillCurvesData( curves );
+		}
+		else if( const auto *points = runTimeCast<const PointsPrimitive>( primitive ) )
+		{
+			fillPointsData( points );
+		}
 	}
 
 	~Expandulator() = default;
 
-	ConstV3fVectorDataPtr expandulateVertex( const PrimitiveVariable::IndexedView<Imath::V3f> &in )
+	/// \todo: should this return a MVertexBuffer directly?
+	ConstV3fVectorDataPtr expandulateVertex( const PrimitiveVariable::IndexedView<Imath::V3f> &in ) const
```

Comment:
I've fixed the indentation issue (I assume you were talking about the non-dangling `(`)

> this whole file is full of parallel_for doing incredibly tiny amounts of work, but with a default grain size ( which I think just means 1? ). Such tiny bits of work are well outside the usual use case for parallel_for, and I rather expect that this could be suffering from a whole lot of overhead. Increasing the grain size might alleviate this - if parallel_for is even appropriate 

Sorry, this is one of the things that is hard to tell from my single large commit -- none of that is a change from master. I intentionally did not touch any "how to compute something" algorithms (other than `fillPoints()` where I now compute indices because otherwise it crashes).

I'll take a guess that they do matter though. Its not "normal"  to write `parallel_for` loops without first measuring the need for it and I know previously authors of this code did quite a lot of vtune profiling. So I'm assuming those 5 loops were all the result of profiling on production use cases (presumably on some incredibly dense lidar meshes).